### PR TITLE
[query] Remove invalid boundary on RepartitionedOrderedRDD2

### PR DESCRIPTION
**Pull request by tpoterba**
Fixes TableMapPartitions(indexed read)

### CODE REVIEW

**Comment by johnc1231** [comment]
Something is wrong here, you have tons of failing python tests.

**Comment by tpoterba** [comment]
yeah... not looking forward to this debugging.

**Comment by tpoterba** [comment]
I think this was probably a "two wrongs *did* make a right" scenario

**Comment by johnc1231** [comment]
@patrick-schultz is usually good at seeing when we are breaking region memory management rules. But yeah, this seems like it broke everything.

**Comment by patrick-schultz** [comment]
Could you explain why you think the boundary is invalid?

**Comment by patrick-schultz** [comment]
Hmm, I think boundary is just unsound in general. It doesn't ensure that the pointers returned by `next` live as long as the consumer region.

I'll look over the PR and see if I can find any problems.

**Comment by tpoterba** [comment]
> Could you explain why you think the boundary is invalid?

Boundary can only be used by consumers who generate an iterator with a new context. Here we were just inserting a clear before each next(), no matter who was consuming the iterator. The particular pipeline that triggered this error was a TableMapPartitions with a ToArray(Ref rows) after an IR with a repartitionedOrderedRDD2

File: hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala [hunk 1]
```diff
@@ -55,20 +55,74 @@ class RepartitionedOrderedRDD2 private (@transient val prev: RVD, @transient val
...
-        }.takeWhile { ptr =>
-          ur.set(ctx.r, ptr)
-          pord.lteq(key, range.right)
+
+    Iterator.single { (outerCtx: RVDContext) =>
+      new Iterator[Long] {
+        private[this] val innerCtx = outerCtx.freshContext()
+        private[this] val outerRegion = outerCtx.region
+        private[this] val innerRegion = innerCtx.region
+        private[this] val parentIterator = ordPartition.parents.iterator.flatMap(p => prevCRDD.iterator(p, context).next().apply(innerCtx))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:64]: You changed from `flatMap` to `next`. I'm not sure we can assume a ContextRDD has exactly one element per partition.

File: hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala [hunk 2]
```diff
@@ -55,20 +55,74 @@ class RepartitionedOrderedRDD2 private (@transient val prev: RVD, @transient val
...
+        private[this] val ur = new UnsafeRow(typ.rowType)
+        private[this] val key = new SelectFieldsRow(ur, typ.kFieldIdx)
+
+        // drop left elements at iterator allocation to avoid extra control flow in hasNext()
+        dropLeft()
+
+        private[this] def dropLeft(): Unit = {
+          var eltLessThan = true
+          while (parentIterator.hasNext && eltLessThan) {
+            innerRegion.clear()
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:76]: This looks fishy. You're clearing the producer region after calling `hasNext` but before calling `next`. The `parentIterator` may have produced the element when you called `hasNext`, then you clear the region, then `next` returns the dangling pointer.

One way to fix it:
```scala
@tailrec def dropLeft(): Unit = {
  if (parentIterator.hasNext) {
    pull()
    if (pord.lt(key, range.left) {
      innerRegion.clear()
      dropLeft()
    } else if (pord.gt(key, range.right) {
      end()
    }
  } else {
    end()
  }
}
```

**Review by patrick-schultz — DISMISSED** [review]

File: hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:64]: really? oh, I'll change that.

File: hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:76]: ahhhhh, yeah, this has to be it. Thanks Patrick!

Comment (tpoterba) [reply hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala:76]: one of the failing tests is passing, I'll push it up! 🎉

Comment (johnc1231) [reply hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala]: Thanks for catching that Patrick.

**Comment by johnc1231** [comment]
This looks good to me now.

**Review by johnc1231 — APPROVED** [review]

**Comment by johnc1231** [comment]
Thanks for tracking this down Tim

**Pull request merged**